server/src/main/java/org/elasticsearch/common/time/DateUtils.java

```diff
@@ -293,6 +293,17 @@ public static long toMilliSeconds(long nanoSecondsSinceEpoch) {
         return nanoSecondsSinceEpoch / 1_000_000;
     }
 
+    public static int compareNanosToMillis(long nanos, long millis) {
+        if (millis < 0) {
```

Comment:
If this is sub-second duration (see question above), how can millis be negative? If it's epoch times, why would negative millis the nanos are always greater?

Reply:
Elasticsearch doesn't allow nanosecond resolution dates before epoch, so a millisecond date before epoch will, by definition, be before any valid nanosecond date. See, for example, `clampToNanosRange`.

Reply:
We didn't support passing negative epoch millis until recently. Rather than make this assumption, why not make the method correctly compare the two?

Reply:
I didn't do that because we then need to deal with the long overflow case in the comparison.  Right now, `nanos - (millis * 1_000_000)` will never overflow, but if we allow negative values there, then overflow becomes a possibility.  We can deal with that, but this seemed more straight forward.  I've added some test cases and comments to make this more explicit.

The assumption that nanosecond dates will never be negative is fairly deeply baked into the system, and we would need to change a lot of things before we could support that, not just this method.  I could change this to read `if (millis < 0 && nanos >= 0) {`, even though the second clause will always be true, if you want.  I'd rather not add further logic for dealing with currently impossible negative nanos cases though, since none of the other tooling supports it and I'd have to hand-roll conversions (and difficulty in those conversions was why we decided not to support pre-epoch nanos in the first place, IIRC).

Reply:
good idea.  Done.